cpp/src/arrow/util/byte_stream_split.h

```diff
@@ -568,45 +574,141 @@ template <typename T>
 void inline ByteStreamSplitEncodeSimd(const uint8_t* raw_values, const size_t num_values,
                                       uint8_t* output_buffer_raw) {
 #if defined(ARROW_HAVE_AVX512)
-  return ByteStreamSplitEncodeAvx512<T>(raw_values, num_values, output_buffer_raw);
+  return ByteStreamSplitEncodeAvx512<T>(raw_values, static_cast<size_t>(num_values),
+                                        output_buffer_raw);
 #elif defined(ARROW_HAVE_AVX2)
-  return ByteStreamSplitEncodeAvx2<T>(raw_values, num_values, output_buffer_raw);
+  return ByteStreamSplitEncodeAvx2<T>(raw_values, static_cast<size_t>(num_values),
+                                      output_buffer_raw);
 #elif defined(ARROW_HAVE_SSE4_2)
-  return ByteStreamSplitEncodeSse2<T>(raw_values, num_values, output_buffer_raw);
+  return ByteStreamSplitEncodeSse2<T>(raw_values, static_cast<size_t>(num_values),
+                                      output_buffer_raw);
 #else
 #error "ByteStreamSplitEncodeSimd not implemented"
 #endif
 }
 #endif
 
+//
+// Scalar implementations
+//
+
+// CAUTION when editing this: on x86 platforms, the test suite only exercises
+// the SIMD versions by default. Either check on non-x86, or manually disable
+// SIMD in the ByteStreamSplitEncode/ByteStreamSplitDecode functions below.
+
+inline void DoSplitStreams(const uint8_t* src, int width, int64_t nvalues,
+                           uint8_t** dest_streams) {
+  // Value empirically chosen to provide the best performance on the author's machine
+  constexpr int kBlockSize = 32;
+
+  while (nvalues >= kBlockSize) {
+    for (int stream = 0; stream < width; ++stream) {
+      uint8_t* dest = dest_streams[stream];
+      for (int i = 0; i < kBlockSize; i += 8) {
+        uint64_t a = src[stream + i * width];
+        uint64_t b = src[stream + (i + 1) * width];
+        uint64_t c = src[stream + (i + 2) * width];
+        uint64_t d = src[stream + (i + 3) * width];
+        uint64_t e = src[stream + (i + 4) * width];
+        uint64_t f = src[stream + (i + 5) * width];
+        uint64_t g = src[stream + (i + 6) * width];
+        uint64_t h = src[stream + (i + 7) * width];
+#if ARROW_LITTLE_ENDIAN
```

Comment:
What if the endianness of machine compiling it is different with machine at runtime?

Reply:
Then the whole of Arrow C++ is miscompiled?

Reply:
Is there any situation where compile-time `ARROW_LITTLE_ENDIAN` might be different from the runtime platform's endianness?

Reply:
Cross-compiling?

Reply:
Here is the current definition. Can it be mis-compiled?
```cpp
#ifndef __BYTE_ORDER__
#error "__BYTE_ORDER__ not defined"
#endif
#
#ifndef __ORDER_LITTLE_ENDIAN__
#error "__ORDER_LITTLE_ENDIAN__ not defined"
#endif
#
#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
#define ARROW_LITTLE_ENDIAN 1
#else
#define ARROW_LITTLE_ENDIAN 0
#endif
```

Reply:
This may work in most of time. Unless we switch to std::endian in C++20 to check it at runtime, depending on predefined macros at compile time is not always reliable: https://stackoverflow.com/questions/8978935/detecting-endianness

Reply:
You may want to open an issue for this. There doesn't seem to be a reliable way to make a compile-time or `constexpr` check in C++17, AFAICT.

Reply:
> Then the whole of Arrow C++ is miscompiled?

Maybe it's ok to test it when starting the user-side program. Since this is not the only place may causing problem when cross-compiling worked.

Reply:
What about opening an issue to switch to std::endian once we start C++20 adoption?

Reply:
I'm not familiar with this part of code, however, that maybe means introduce a kind of dynamic code dispatch, like if LITTLE_ENDIAN , otherwise goes to BIG_ENDIAN code.

As you referenced:

> As stated earlier, the only "real" way to detect Big Endian is to use runtime tests.

`std::endian` seems also do a compile time checking rather than runtime. See https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/p0463r1.html

Reply:
IMO, a lots of code already uses compile-time macro. If we meet this problem, we may having a test for SIMD instructions, endian, etc. Rather than dynamic dispatch it here...